<a href="https://colab.research.google.com/github/serialize02/2023final/blob/main/classification_ipynb_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Loading a dataset

In [1]:
import pandas as pd

### iris dataset

In [2]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
iris = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data', names=['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'label'])
iris['label'] = le.fit_transform(iris['label'])
iris.head()

sepal_length  sepal_width  petal_length  petal_width  label
0           5.1          3.5           1.4          0.2      0
1           4.9          3.0           1.4          0.2      0
2           4.7          3.2           1.3          0.2      0
3           4.6          3.1           1.5          0.2      0
4           5.0          3.6           1.4          0.2      0

4종류의 숫자는 꽃에서 어떤 부위를 measure 한 것
sepal - 꽃받침, petal-꽃잎

처음에 받아오면 header 가 없어서 헤더도 동시에 부여를 해준다.
마지막은 label 꽃의 종류 이름


In [3]:
iris

sepal_length  sepal_width  petal_length  petal_width  label
0             5.1          3.5           1.4          0.2      0
1             4.9          3.0           1.4          0.2      0
2             4.7          3.2           1.3          0.2      0
3             4.6          3.1           1.5          0.2      0
4             5.0          3.6           1.4          0.2      0
..            ...          ...           ...          ...    ...
145           6.7          3.0           5.2          2.3      2
146           6.3          2.5           5.0          1.9      2
147           6.5          3.0           5.2          2.0      2
148           6.2          3.4           5.4          2.3      2
149           5.9          3.0           5.1          1.8      2

[150 rows x 5 columns]

총 150개의 붓꽃, 종류 3가지가 50개씩
이 데이터의 목적은?? -> 내가 꽃의 measure 데이터를 넣었을 때, 이건 무슨 꽃이다 라고 맞춰주는 것이 목표

### breast cancer dataset

실질적으로 breast cancer 인지 아닌지만 알면 됨
regression의 값이 아니라 discrete 한 output

데이터에서 필요 없는 부분을 좀 정제해주는 과정이 필요한데,
id 부분은 필요 없으니까 삭제, 2,4를 0,1로 변환, 그리고 물음표를 집어넣으면 에러가 나기 때문에 이걸 아주 큰 값으로 넣어서 알아서 해결하게 함

In [4]:
breast_cancer = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/breast-cancer-wisconsin/breast-cancer-wisconsin.data', names=['id', 'clump_thickness', 'uniformity_of_cell_size', 'uniformity_of_cell_shape', 'marginal_adhesion', 'single_epithelial_cell_size', 'bare_nuclei', 'bland_chromatin', 'normal_nucleoli', 'mitoses', 'label'])
breast_cancer.drop(['id'], axis=1, inplace=True)
breast_cancer['label'].replace([2, 4], [0, 1], inplace=True)
breast_cancer.replace('?', -999999, inplace=True)  # replace missing value with outliers
breast_cancer.head()

clump_thickness  uniformity_of_cell_size  uniformity_of_cell_shape  \
0                5                        1                         1   
1                5                        4                         4   
2                3                        1                         1   
3                6                        8                         8   
4                4                        1                         1   

   marginal_adhesion  single_epithelial_cell_size bare_nuclei  \
0                  1                            2           1   
1                  5                            7          10   
2                  1                            2           2   
3                  1                            3           4   
4                  3                            2           1   

   bland_chromatin  normal_nucleoli  mitoses  label  
0                3                1        1      0  
1                3                2        1      0  
2                3                1        1      0  
3                3                7        1      0  
4                3                1        1      0

### wine dataset

In [5]:
wine = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/wine/wine.data', names=['label', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium', 'total_phenols','flavanoids', 'nonflavanoid_phenols' ,'proanthocyanins', 'color_intensity', 'hue', 'OD280/OD315_of_diluted_wines', 'proline'], delimiter=",", index_col=False)
wine.head()

<ipython-input-5-07571967278c>:1: ParserWarning: Length of header or names does not match length of data. This leads to a loss of data with index_col=False.
  wine = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/wine/wine.data', names=['label', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium', 'total_phenols','flavanoids', 'nonflavanoid_phenols' ,'proanthocyanins', 'color_intensity', 'hue', 'OD280/OD315_of_diluted_wines', 'proline'], delimiter=",", index_col=False)


label  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      1       14.23  1.71               2.43       15.6            127   
1      1       13.20  1.78               2.14       11.2            100   
2      1       13.16  2.36               2.67       18.6            101   
3      1       14.37  1.95               2.50       16.8            113   
4      1       13.24  2.59               2.87       21.0            118   

   flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0        2.80                  3.06             0.28             2.29  5.64   
1        2.65                  2.76             0.26             1.28  4.38   
2        2.80                  3.24             0.30             2.81  5.68   
3        3.85                  3.49             0.24             2.18  7.80   
4        2.80                  2.69             0.39             1.82  4.32   

   OD280/OD315_of_diluted_wines  proline  
0                          1.04     3.92  
1                          1.05     3.40  
2                          1.03     3.17  
3                          0.86     3.45  
4                          1.04     2.93

## Comparing 6 classification algorithms

In [6]:
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neural_network import MLPClassifier

models = [
    ('LR', LogisticRegression()),
    ('NB', GaussianNB()),
    ('SVM', SVC()),
    ('KNN', KNeighborsClassifier()),
    ('DT', DecisionTreeClassifier()),
    ('NN', MLPClassifier()),
]

이것들이 다 classification 관련된 알고리즘들임
대부분 뉴럴넷 -mlp classifier 이걸 많이 씀 (챗지피티도 이거임)
위에것들은 조금 옛날거고 이제는 나름 통일이 됐구나 알면 됨.

### test a single model (e.g. Logistic Regression)

In [7]:
from sklearn.model_selection import train_test_split
import numpy as np

dataset_name = 'iris'
dataset = iris

X = np.array(dataset.drop(['label'], axis=1))
y = np.array(dataset['label'])
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

name = 'LR'
model = LogisticRegression()
model.fit(X_train, y_train)
accuracy = model.score(X_test, y_test)
print(dataset_name, name, accuracy)

iris LR 1.0


어떤 알고리즘을 사용하든, 입력부에 들어가는 것을 먼저 정해야함.
이 경우, 숫자 4개에 label 1개의 데이터가 있는데, 숫자 4개가 입력부 쪽, 데이터 1개가 출력부 쪽에 배치되어야 함

그래서 입력부 부분에서 label 부분을 제외해야 함 -> 그거를 array 로 만들어라
출력부는? label 만 들고가라

입력의 데이터와 출력의 데이터를 엑셀 파일같은 데이터 프레임에서 쪼개야 함

그리고 또 실제 train 용 할 때 일부분만 씀
테스트 과정이기 때문에 일부 남겨놓고 나중에 잘됐는지 안됐는지는 실험 용으로 사용
-> 두번의 쪼개는 과정이 필요
입력용/출력용 , 훈련용/테스트용

model.fit .fit 는 이렇게 훈련을 해라 ~ 라는 말임
실질적으로 훈련용만 필요하니까 거기만 가져와서 하면 됨

model.score = score, accuracy, performance, test 다 그냥 같은 말로 생각

이 경우 100%의 정확도라고 나옴

In [8]:
model.predict(np.array([6, 3, 5, 2]).reshape(1,4))

array([2])

predict

저런저런 정보를 줬더니 ~ 3번째 꽃이다 !!(0,1,2)
이 자체로 쓸 수 있으니 응용이 가능한 상태가 됨

### Evaluate all the models

데이터도 3가지고, 우리가 쓸 수 있는 알고리즘도 여러가지인데, 한꺼번에 어떤게 더 좋은가? 를 비교하는 방법

하나 당 루프 2개가 돌아가는데, 이걸 3개의 데이터에 대해서 6개의 알고리즘으로 하니까 총 18개의 accuracy, performance가 나오는 것

In [9]:
for dataset_name, dataset in [('iris', iris), ('breast_cancer', breast_cancer), ('wine', wine)]:
  X = np.array(dataset.drop(['label'], axis=1))
  y = np.array(dataset['label'])
  X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

  for name, model in models:
    model.fit(X_train, y_train)
    accuracy = model.score(X_test, y_test)
    print(dataset_name, name, accuracy)

iris LR 1.0
iris NB 1.0
iris SVM 1.0
iris KNN 1.0
iris DT 1.0
iris NN 1.0
breast_cancer LR 0.35714285714285715
breast_cancer NB 0.9214285714285714
breast_cancer SVM 0.6785714285714286
breast_cancer KNN 0.9857142857142858
breast_cancer DT 0.95


/usr/local/lib/python3.10/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


breast_cancer NN 0.95
wine LR 0.9166666666666666
wine NB 0.9444444444444444
wine SVM 0.5833333333333334
wine KNN 0.8333333333333334
wine DT 0.9722222222222222
wine NN 0.8888888888888888


/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.10/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


데이터별로 어떤 알고리즘 모델을 선택하는게 좋은지 정확도를 기반으로 판단하면 됨

시험문제
하나의 데이터가 있을 때 이것은 반드시 classification 과 regression 둘 중 하나만으로 쓸 수 있다. 틀림.